### PR TITLE
implement grep in go

### LINKED ISSUE #1: Implement grep in Go
... so that `go run . -e 'fo*bar' < in.txt` does the same as `grep -e 'fo*bar' < in.txt`.

`fo*bar` may be Go regex instead of grep's standard regex.

### PR DESCRIPTION
**Pull request by justinmueller05082000**
fixes #1

### CODE REVIEW

**Comment by Al2Klimov** [comment]
"refs"?

File: main.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+			if re.Match(data) {
+				_, _ = os.Stdout.Write(data)
+			}
+
+			if dataErr == io.EOF {
+				break
+			}
+		}
+		return
+	}
```
Comment (Al2Klimov) [inline main.go]: -e is a required parameter. How did you handle missing required parameters before?

File: main.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+			if re.Match(data) {
+				_, _ = os.Stdout.Write(data)
+			}
+
+			if dataErr == io.EOF {
+				break
+			}
+		}
+		return
```
Comment (Al2Klimov) [inline main.go]: What's this for?

File: main.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package main
```
Comment (Al2Klimov) [inline main.go:3]: What's missing here?

File: main.go [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+			if dataErr != nil && dataErr != io.EOF {
+				fmt.Fprintln(os.Stderr, dataErr)
+				os.Exit(1)
+			}
+
+			if len(data) == 0 {
+				break
+			}
+
+			if data[len(data) - 1] != '\n' {
```
Comment (Al2Klimov) [inline main.go]: Isn't there already a function for checking this?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
What happens on `go run . -e '*'?`

File: main.go (hunk 3 above)
Comment (justinmueller05082000) [reply main.go:3]: license header?

Comment (Al2Klimov) [reply main.go:3]: Exactly.

File: main.go (hunk 1 above)
Comment (justinmueller05082000) [reply main.go]: i print an error message and exit the program?

Comment (Al2Klimov) [reply main.go]: Exactly.

**Comment by justinmueller05082000** [comment]
Screenshot from 2020-06-03 10-30-32
This happens

**Comment by Al2Klimov** [comment]
Make it more like an error message.

**Comment by justinmueller05082000** [comment]
So should i just check if my parameter is '*' and if i print out an error message?

**Comment by Al2Klimov** [comment]
No, I demand a more generic solution.

**Comment by justinmueller05082000** [comment]
but what should i check for? I cannot check for the length because u can also grep with just one character i guess like 'a'

**Comment by Al2Klimov** [comment]
**You** don't check the regex validity. **A function** which converts string to regex checks it and tells you as error. But the function you use at the moment panics instead of telling you.

File: main.go [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+			if re.Match(data) {
+				_, _ = os.Stdout.Write(data)
+			}
+
+			if dataErr == io.EOF {
+				break
+			}
+		}
+	} else {
```
Comment (Al2Klimov) [inline main.go]: Please put this out of the if (w/o changing the program's behavior) as in the other programs.

File: main.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+			if dataErr != nil && dataErr != io.EOF {
+				fmt.Fprintln(os.Stderr, dataErr)
+				os.Exit(1)
+			}
+
+			if len(data) == 0 {
+				break
+			}
+
+			if !bytes.Contains(data, []byte{'\n'}) {
```
Comment (Al2Klimov) [inline main.go]: It works, but isn't the function I meant.

File: main.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+			if re.Match(data) {
+				_, _ = os.Stdout.Write(data)
+			}
+
+			if dataErr == io.EOF {
+				break
+			}
+		}
+	} else {
+		fmt.Fprintln(os.Stderr, "grep: at least one parameter is needed.")
```
Comment (Al2Klimov) [inline main.go]: "at least one" of **which** parameter? Yes, at the moment there's only one, but have a look at the issue list.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 7 above)
Comment (justinmueller05082000) [reply main.go]: you can see the parameters with grep --help or in my case go run . --help

Comment (Al2Klimov) [reply main.go]: Yes, I see there **what's available**, but in the error message I don't see **what I missed**.

File: main.go (hunk 6 above)
Comment (justinmueller05082000) [reply main.go]: I dont know which function you mean then

Comment (Al2Klimov) [reply main.go]: The one you've already used in a previous program.

File: main.go (hunk 7 above)
Comment (Al2Klimov) [reply main.go]: The message was almost right, you had just to add that it's about the **-e** parameter.

File: main.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+			os.Exit(1)
+		}
+
+		if len(data) == 0 {
+			break
+		}
+
+		if !bytes.HasSuffix(data, []byte{'\n'}) {
+			data = append(data, '\n')
+		}
```
Comment (Al2Klimov) [inline main.go]: Do you always need this block? Or just on certain conditions?

**Review by Al2Klimov — CHANGES_REQUESTED** [review]

File: main.go (hunk 8 above)
Comment (justinmueller05082000) [reply main.go]: i need to check every single line so i think I need it always

Comment (Al2Klimov) [reply main.go]: The block eventually modifies `data`. Do you always need `data` modified this way? Or only if... ?

Comment (justinmueller05082000) [reply main.go]: data is a line of a text file and if the line has no newline at the end I add one and if not then not. I dont get your point

Comment (Al2Klimov) [reply main.go]: **What exactly** do you need the newline for?

Comment (justinmueller05082000) [reply main.go]: for my output to be equal to greps output

Comment (Al2Klimov) [reply main.go]: Right! Your **output**! But you don't output every line, do you?

Comment (justinmueller05082000) [reply main.go]: I output the line where my regex expression is found in? do you want me to put this code into my if re.match?

Comment (Al2Klimov) [reply main.go]: Exactly.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
* [x] `git reset --soft HEAD~$(git rev-list --count $(git merge-base HEAD master)..HEAD^)`
* [x] `git commit --amend --no-edit`
* [x] `git push -f origin feature/implement-grep-in-go-1`

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol.

**Comment by justinmueller05082000** [comment]
**Test protocol:**

**toGrep.txt:**

```
foobar dog cat cat foobar dog cat mouse mouse
tiger dog cat dog mouse mouse zebra elephant
dog cat foobar mouse zebra zebra cat dog
tiger dog cat mouse dog dog cat zebra zebra tiger
elephant foobar foobar elephant foobar
```

**Go code:**

**Terminal:** `go run . -e  'fo*bar' < toGrep.txt`

**Output:**

```
foobar dog cat cat foobar dog cat mouse mouse
dog cat foobar mouse zebra zebra cat dog
elephant foobar foobar elephant foobar
```

**Grep command:**

**Terminal:**  `grep -e 'fo*bar' < toGrep.txt`

**Output:**

```
foobar dog cat cat foobar dog cat mouse mouse
dog cat foobar mouse zebra zebra cat dog
elephant foobar foobar elephant foobar
```


**Conclusion:**
Both results are equal except for the colored highlighting in the terminal.

**Review by Al2Klimov — CHANGES_REQUESTED** [review]
Write a test protocol testing a multi-line file, proving that your program actually filters lines.

**Review by Al2Klimov — APPROVED** [review]

**Pull request merged**